utils_nlp/models/transformers/extractive_summarization.py

```diff
@@ -26,6 +28,75 @@
 
 logger = logging.getLogger(__name__)
 
+# https://pytorch.org/docs/1.1.0/_modules/torch/utils/data/dataloader.html
+class IterableDistributedSampler(object):
+    """ Distributed sampler for iterable dataset.
+
+    Args:
+        world_size (int): Total number of GPUs that will be used. Defaults to 1.
+        rank (int): Rank of the current GPU. Defaults to 1.
+        
+    """
+    def __init__(self, world_size=1, rank=-1):
+        self.world_size = world_size
+        self.rank = rank
+
+    def iter(self, iterable):
+        if self.rank != -1:
+            return itertools.islice(iterable, self.rank, None, self.world_size)
+        else:
+            return iterable
+
+class ChunkDataLoader(object):
+    """ Data Loader for Chunked Dataset.
+
+    Args:
+        datasets (list): list of data item list.
+        batch_size (int): Number of tokens per batch.
+        shuffle (bool): Whether the data is shuffled.
+        is_labeled (bool): Whether the data is labeled.
+        sampler (obj): Data sampler.
+
+    """
+    def __init__(self, datasets,  batch_size,
+                 shuffle, is_labeled, sampler):
+        self.datasets = datasets
+        self.batch_size = batch_size
+        self.shuffle = shuffle
+        self.is_labeled = is_labeled
+        self.cur_iter = self._next_dataset_iterator(datasets)
+        assert self.cur_iter is not None
+        self.sampler = sampler
+
+
+    def eachiter(self):
+        dataset_iter = (d for d in self.datasets)
+        while self.cur_iter is not None:
+            for batch in self.cur_iter:
+                yield batch
+            self.cur_iter = self._next_dataset_iterator(dataset_iter)
+
+    def __iter__(self):
+        return self.sampler.iter(self.eachiter())
+
+
+    def _next_dataset_iterator(self, dataset_iter):
+        try:
+            # Drop the current dataset for decreasing memory
+            if hasattr(self, "cur_dataset"):
+                self.cur_dataset = None
```

Comment:
Curious why `del self.cur_dataset` and `gc.collect()` is not enough. I guess there must be a reason.

Reply:
Not sure whether it's enough or not. The code is directly copied from the original source.

Reply:
my guess is that line 87 makes sure the object that self.cur_dataset is de-referenced. and the "del" is used to decrease the reference counter. Not sure why "del" needs to be called twice.